### PR TITLE
[Brave Leo]: Add Claude 3 client side model support

### PR DESCRIPTION
**Pull request by LorenzoMinto**
Resolves https://github.com/brave/brave-browser/issues/36802
Depends on https://github.com/brave/aichat/pull/175

## Submitter Checklist:

- [ ] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/ai_chat/core/browser/models.cc [hunk 1]
```diff
@@ -43,9 +43,15 @@ const std::vector<ai_chat::mojom::Model>& GetAllModels() {
...
-      {"chat-claude-instant", "claude-instant-v1", "Claude Instant",
+       {"chat-claude-instant", "claude-instant-v1", "Claude Instant",
        "Anthropic", mojom::ModelEngineType::CLAUDE_REMOTE,
        mojom::ModelCategory::CHAT, kFreemiumAccess, 180000, 320000},
+       {"chat-claude-haiku", "claude-3-haiku-20240307", "Claude 3 Haiku",
+       "Anthropic", mojom::ModelEngineType::CLAUDE_REMOTE,
+       mojom::ModelCategory::CHAT, kFreemiumAccess, 180000, 320000},
+       {"chat-claude-sonnet", "claude-3-sonnet-20240229", "Claude 3 Sonnet",
+       "Anthropic", mojom::ModelEngineType::CLAUDE_REMOTE,
+       mojom::ModelCategory::CHAT, mojom::ModelAccess::PREMIUM, 180000, 320000},
```
[the remark this answers is not in the document]
Comment (LorenzoMinto) [reply components/ai_chat/core/browser/models.cc:54]: long term code names 👍 I guess it's part of the branding of this model family

[the remark this answers is not in the document]
Comment (stephendonner) [reply components/ai_chat/core/browser/models.cc:54]: not to be sacrilegious but it's like the holy/divine trinity from the Judeo-Christian Bible

File: browser/ai_chat/ai_chat_settings_helper.cc [hunk 2]
```diff
@@ -81,6 +81,14 @@ void AIChatSettingsHelper::GetModelsWithSubtitles(
...
           l10n_util::GetStringUTF8(IDS_CHAT_UI_CHAT_CLAUDE_INSTANT_SUBTITLE);
       is_key_handled = true;
+    } else if (modelWithSubtitle.model->key == "chat-claude-haiku") {
+      modelWithSubtitle.subtitle =
+          l10n_util::GetStringUTF8(IDS_CHAT_UI_CHAT_CLAUDE_HAIKU_SUBTITLE);
+      is_key_handled = true;
+    } else if (modelWithSubtitle.model->key == "chat-claude-sonnet") {
+      modelWithSubtitle.subtitle =
+          l10n_util::GetStringUTF8(IDS_CHAT_UI_CHAT_CLAUDE_SONNET_SUBTITLE);
+      is_key_handled = true;
```
Comment (petemill) [inline browser/ai_chat/ai_chat_settings_helper.cc:91]: cc for @Brandon-T as I don't think this class / function with the model descriptions is shared with iOS yet, unfortunately.

**Review by petemill — APPROVED** [review]

File: browser/ai_chat/ai_chat_settings_helper.cc (hunk 2 above)
Comment (Brandon-T) [reply browser/ai_chat/ai_chat_settings_helper.cc:91]: Yes. iOS has to add these manually. We don't use the WebUI settings screen. I might be able to refactor this in the future to pull in the info and they'd both share the underlying code.

But for now, iOS has to add this.

Comment (LorenzoMinto) [reply browser/ai_chat/ai_chat_settings_helper.cc:91]: Should I add it to iOS or open an issue for a followup?

**Review by Brandon-T — APPROVED** [review]

**Comment by stephendonner** [comment]
## Verification `IN-PROGRESS` using

```
Brave | 1.67.73 Chromium: 125.0.6422.26 (Official Build) nightly (64-bit)
-- | --
Revision | 76e7d97680074902ad7bde84f0a1a6c392a4ca6a
OS | Windows 10 Version 22H2 (Build 19045.4355)
```

**Pull request merged**